src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAIBasic.java

```diff
@@ -606,344 +610,164 @@ public void takeItemStackFromProvider(@NotNull final ICapabilityProvider provide
+               ItemStackUtils.swapToolGrade(maximumLevel));
 
-    /**
-     * Looks for a pickaxe to mine a block of {@code minLevel}.
-     * The pickaxe will be taken from the chest.
-     * Make sure that the worker stands next the chest to not break immersion.
-     * Also make sure to have inventory space for the pickaxe.
-     *
-     * @param minlevel the needed pickaxe level
-     * @return true if a pickaxe was found
-     */
-    private boolean isPickaxeInHut(final int minlevel)
-    {
-        @Nullable final AbstractBuildingWorker buildingWorker = getOwnBuilding();
-        return buildingWorker != null
-                && InventoryFunctions.matchFirstInProviderWithAction(
-          buildingWorker.getTileEntity(),
-          stack -> Utils.checkIfPickaxeQualifies(
-            minlevel,
-            Utils.getMiningLevel(stack, Utils.PICKAXE)
-          ),
-                this::takeItemStackFromProvider
-        );
-    }
-
-    /**
-     * Wait for a needed pickaxe.
-     *
-     * @return NEEDS_PICKAXE
-     */
-    @NotNull
-    private AIState waitForWeapon()
-    {
-        if (checkForWeapon())
-        {
-            delay += DELAY_RECHECK;
-            return NEEDS_WEAPON;
         }
-        return IDLE;
     }
 
     /**
-     * Ensures that we have a pickaxe available.
-     * Will set {@code needsPickaxe} accordingly.
+     * Check all chests in the worker hut for a required tool.
      *
-     * @return true if we have a pickaxe
+     * @param toolType the type of tool requested (amount is ignored)
+     * @param minimalLevel the minimal level the tool should have.
+     * @return true if a stack of that type was found
      */
-    public boolean checkForWeapon()
+    public boolean isToolInHut(final IToolType toolType, final int minimalLevel)
     {
-        //Check for a pickaxe
-        getOwnBuilding().setNeedsWeapon(
-                !InventoryFunctions.matchFirstInProvider(
-                        worker,
-                        stack -> stack != null && Utils.doesItemServeAsWeapon(stack)));
-
-        delay += DELAY_RECHECK;
+        @Nullable final AbstractBuildingWorker building = getOwnBuilding();
 
-        if (getOwnBuilding().needsWeapon())
+        if (building != null)
         {
-            if (walkToBuilding())
+            if (isToolInTileEntity(building.getTileEntity(), toolType, minimalLevel, getOwnBuilding().getMaxToolLevel()))
             {
-                return false;
+                return true;
             }
-            if (isWeaponInHut())
+
+            for (final BlockPos pos : building.getAdditionalCountainers())
             {
-                return true;
+                final TileEntity entity = world.getTileEntity(pos);
+                if (entity instanceof TileEntityChest)
+                {
+                    if (isToolInTileEntity((TileEntityChest) entity, toolType, minimalLevel, getOwnBuilding().getMaxToolLevel()))
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Merge this if statement with the enclosing one. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1066)